src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java

```diff
@@ -4101,13 +4101,17 @@ protected static ClassNode getGroupOperationResultType(ClassNode a, ClassNode b)
         return Number_TYPE;
     }
 
-    protected ClassNode inferComponentType(final ClassNode containerType, final ClassNode indexType) {
+    private static ClassNode convertGStringToStringType(ClassNode cn) {
+        return GSTRING_TYPE.equals(cn) ? STRING_TYPE : cn;
+    }
+
+    protected ClassNode inferComponentType(final ClassNode containerType, ClassNode indexType) {
         final ClassNode componentType = containerType.getComponentType();
         if (componentType == null) {
             // GROOVY-5521
             // try to identify a getAt method
             typeCheckingContext.pushErrorCollector();
-            MethodCallExpression vcall = callX(varX("_hash_", containerType), "getAt", varX("_index_", indexType));
+            MethodCallExpression vcall = callX(varX("_hash_", containerType), "getAt", varX("_index_", convertGStringToStringType(indexType)));
```

Comment:
what if I add a DGM getAt(GString) for Map? This code looks like I would no longer be able to call it

Reply:
Good idea! 
```groovy
    /**
     * Get the value via the string value of the key
     * See GROOVY-6668 and the PR#708 (https://github.com/apache/groovy/pull/708)
     *
     * @param self a map
     * @param key a key
     * @param <V> value type
     * @return the value
     */
    public static <V> V getAt(Map<?, V> self, GString key) {
        if (null == key) {
            return null;
        }

        return self.get(key.toString());
    }
```

but I wonder why we should not treat GString as String?
As a groovy user, I won't write code like `GString str = "${'someString'}"`, and I just write `String str = "${'someString'}"` or  `def str = "${'someString'}"` then treat `str` as a `String`...

Reply:
we can treat GString as a String if we know the target is a String. Plus the rules for assignment and for method calls are not identical. But my comment was not about adding getAt(GString) for Map. My comment was that this change looks like the following will no longer work:
  
      class X {
        int getAt(GString gs) {1}
      }
      assert new X()["${42}"] == 1